### PR TITLE
Simulcast conference api

### PR DESCRIPTION
**Pull request by pe1ros**

### CODE REVIEW

File: ios/RNVICallModule.m [hunk 1]
```diff
@@ -158,6 +164,52 @@ @implementation RNVICallModule
...
 }
 
+RCT_EXPORT_METHOD(startReceiving:(NSString *)streamId
+                        resolver:(RCTPromiseResolveBlock)resolve
+                        rejecter:(RCTPromiseRejectBlock)reject) {
+    VIRemoteVideoStream *remoteVideoStream = [RNVICallManager getRemoteVideoStreamById:streamId];
+    if (remoteVideoStream) {
+        [remoteVideoStream startReceivingWithCompletion:^(NSError * _Nullable error) {
+            if (error) {
+                reject(nil, error, nil);
```
Comment (YuliaGrigorieva) [inline ios/RNVICallModule.m:174]: error code should not be null

the same applies to `stopReceiving`, `requestVideoSize` methods

File: ios/RNVIClientModule.m [hunk 2]
```diff
@@ -291,23 +283,21 @@ - (void)stopObserving {
...
-        callSettings.customData = customData;
-        callSettings.extraHeaders = headers;
+        callSettings.customData = [settings valueForKey:@"customData"];
+        callSettings.extraHeaders = [settings valueForKey:@"extraHeaders"];;
+        NSDictionary *videoFlags = [settings valueForKey:@"video"];
         callSettings.videoFlags = [VIVideoFlags videoFlagsWithReceiveVideo:[[videoFlags valueForKey:@"receiveVideo"] boolValue]
                                                                  sendVideo:[[videoFlags valueForKey:@"sendVideo"] boolValue]];
-        callSettings.preferredVideoCodec = videoCodec;
+        callSettings.preferredVideoCodec = [self convertVideoCodecFromString:[settings valueForKey:@"preferredVideoCodec"]];
+        callSettings.enableSimulcast = [settings valueForKey:@"enableSimulcast"];
```
Comment (YuliaGrigorieva) [inline ios/RNVIClientModule.m:297]: this piece of code can be moved to a private method and reused for `callConference` and `answerCall`

File: src/call/Endpoint.js [hunk 3]
```diff
@@ -111,6 +111,35 @@ export default class Endpoint {
         }
     }
 
+    /**
+     * Method for requests the specified video size for the video stream.
+     * @param {*} streamId 
+     * @param {*} width 
+     * @param {*} height 
+     * @returns
```
Comment (YuliaGrigorieva) [inline src/call/Endpoint.js:119]: Please add complete documentation for new APIs

Don't forget to add `@memberOf Voximplant.Endpoint`

**Review by YuliaGrigorieva — CHANGES_REQUESTED** [review]

File: android/src/main/java/com/voximplant/reactnative/CallManager.java [hunk 4]
```diff
@@ -19,7 +20,8 @@
     // endpoint id and call id matching
     private HashMap<String, String> mCallEndpoints = new HashMap<>();
 
-    private HashMap<String, IVideoStream> mVideoStreams = new HashMap<>();
+    private HashMap<String, ILocalVideoStream> localVideoStreams = new HashMap<>();
+    private HashMap<String, IRemoteVideoStream> remoteVideoStreams = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (pe1ros) [reply android/src/main/java/com/voximplant/reactnative/CallManager.java:24]: done

File: android/src/main/java/com/voximplant/reactnative/VICallModule.java [hunk 5]
```diff
@@ -198,6 +201,53 @@ public void hangup(String callId, ReadableMap headers) {
...
+        if (remoteVideoStream != null) {
+            remoteVideoStream.startReceiving(new ICallCompletionHandler() {
+                @Override
+                public void onComplete() {
+                    promise.resolve(null);
+                }
+
+                @Override
+                public void onFailure(CallException exception) {
+                    promise.reject(exception);
```
[the remark this answers is not in the document]
Comment (pe1ros) [reply android/src/main/java/com/voximplant/reactnative/VICallModule.java:216]: done

File: android/src/main/java/com/voximplant/reactnative/VICallModule.java [hunk 6]
```diff
@@ -198,6 +201,53 @@ public void hangup(String callId, ReadableMap headers) {
...
+        if (remoteVideoStream != null) {
+            remoteVideoStream.stopReceiving(new ICallCompletionHandler() {
+                @Override
+                public void onComplete() {
+                    promise.resolve(null);
+                }
+
+                @Override
+                public void onFailure(CallException exception) {
+                    promise.reject(exception);
```
[the remark this answers is not in the document]
Comment (pe1ros) [reply android/src/main/java/com/voximplant/reactnative/VICallModule.java:234]: done

File: android/src/main/java/com/voximplant/reactnative/VICallModule.java [hunk 7]
```diff
@@ -198,6 +201,53 @@ public void hangup(String callId, ReadableMap headers) {
...
+    }
+
+    @ReactMethod
+    public void requestVideoSize(String streamId, int width, int height, final Promise promise) {
+        IRemoteVideoStream remoteVideoStream = CallManager.getInstance().getRemoteVideoStreamById(streamId);
+        if (remoteVideoStream != null) {
+            remoteVideoStream.requestVideoSize(width, height);
+            promise.resolve(null);
+        } else {
+            promise.reject(null, "Can't perform video size for this streamId");
```
[the remark this answers is not in the document]
Comment (pe1ros) [reply android/src/main/java/com/voximplant/reactnative/VICallModule.java:247]: done

File: ios/RNVICallModule.m (hunk 1 above)
Comment (pe1ros) [reply ios/RNVICallModule.m:174]: done

File: android/src/main/java/com/voximplant/reactnative/VICallModule.java [hunk 8]
```diff
@@ -64,14 +66,15 @@ public void internalSetup(String callId) {
     }
 
     @ReactMethod
-    public void answer(String callId, ReadableMap videoSettings, String videoCodec, String customData, ReadableMap headers) {
+    public void answer(String callId, ReadableMap videoSettings, String videoCodec, String customData, ReadableMap headers, boolean enableSimulcast) {
```
[the remark this answers is not in the document]
Comment (pe1ros) [reply android/src/main/java/com/voximplant/reactnative/VICallModule.java:69]: done

File: android/src/main/java/com/voximplant/reactnative/VIClientModule.java [hunk 9]
```diff
@@ -208,13 +208,19 @@ public void handlePushNotification(ReadableMap notification) {
...
+			ReadableMap map = settings.getMap("video");
+			if (map != null) {
+				callSettings.videoFlags = new VideoFlags(map.getBoolean("receiveVideo"),map.getBoolean("sendVideo"));
+			} else {
+				callSettings.videoFlags = new VideoFlags(true, true);
+			}
+			callSettings.customData = settings.getString("customData");
+			callSettings.extraHeaders = Utils.createHashMap(settings.getMap("extraHeaders"));
+			callSettings.preferredVideoCodec = Utils.convertStringToVideoCodec(settings.getString("preferredVideoCodec"));
+			callSettings.enableSimulcast = settings.getBoolean("enableSimulcast");
```
[the remark this answers is not in the document]
Comment (pe1ros) [reply android/src/main/java/com/voximplant/reactnative/VIClientModule.java:223]: done

File: src/call/Endpoint.js (hunk 3 above)
Comment (pe1ros) [reply src/call/Endpoint.js:119]: done

File: src/call/Endpoint.js [hunk 10]
```diff
@@ -111,6 +112,39 @@ export default class Endpoint {
...
     }
 
+    /**
+     * Method for requests the specified video size for the video stream.
+     * @param {*} streamId - Remote video stream id
+     * @param {*} width - Requested width of the video stream
+     * @param {*} height - Requested height of the video stream
+     * @returns {Promise<void|CallError>}
+     * @memberOf Voximplant.Endpoint
+     * @returns
```
Comment (YuliaGrigorieva) [inline src/call/Endpoint.js:122]: extra return

File: src/call/Endpoint.js [hunk 11]
```diff
@@ -111,6 +112,39 @@ export default class Endpoint {
         }
     }
 
+    /**
+     * Method for requests the specified video size for the video stream.
+     * @param {*} streamId - Remote video stream id
```
Comment (YuliaGrigorieva) [inline src/call/Endpoint.js:117]: Please specify types

File: src/client/Client.js [hunk 12]
```diff
@@ -477,8 +479,7 @@ export default class Client {
                     });
             }
             if (Platform.OS === 'ios') {
```
Comment (YuliaGrigorieva) [inline src/client/Client.js:481]: Looks like the same code for both platforms

File: src/client/Client.js [hunk 13]
```diff
@@ -459,15 +459,17 @@ export default class Client {
             callSettings.customData = null;
         }
         if (callSettings.extraHeaders === undefined) {
-            callSettings.extraHeaders = null;
+            callSettings.extraHeaders = {};
         }
         if (callSettings.setupCallKit === undefined || callSettings.setupCallKit === null) {
             callSettings.setupCallKit = false;
         }
```
Comment (YuliaGrigorieva) [inline src/client/Client.js:466]: Please wrap it with Platform.OS === 'ios'

File: src/client/Client.js (hunk 12 above)
Comment (YuliaGrigorieva) [reply src/client/Client.js:481]: the same for `callConference` API

**Review by YuliaGrigorieva — CHANGES_REQUESTED** [review]

File: src/call/Call.js [hunk 14]
```diff
@@ -127,8 +127,11 @@ export default class Call {
         if (callSettings.extraHeaders === undefined) {
             callSettings.extraHeaders = null;
         }
+        if (callSettings.enableSimulcast === undefined) {
+            callSettings.enableSimulcast = false;
+        }
 
-        CallModule.answer(this.callId, callSettings.video, callSettings.preferredVideoCodec, callSettings.customData, callSettings.extraHeaders);
+        CallModule.answer(this.callId, callSettings.video, callSettings.preferredVideoCodec, callSettings.customData, callSettings.extraHeaders, callSettings.enableSimulcast);
```
[the remark this answers is not in the document]
Comment (pe1ros) [reply src/call/Call.js:134]: done

File: src/call/Endpoint.js (hunk 10 above)
Comment (pe1ros) [reply src/call/Endpoint.js:122]: done

File: src/call/Endpoint.js (hunk 11 above)
Comment (pe1ros) [reply src/call/Endpoint.js:117]: done

File: src/client/Client.js (hunk 12 above)
Comment (pe1ros) [reply src/client/Client.js:481]: done

File: src/client/Client.js (hunk 13 above)
Comment (pe1ros) [reply src/client/Client.js:466]: done

**Review by YuliaGrigorieva — APPROVED** [review]

**Pull request merged**